ml/src/main/scala/ai/h2o/sparkling/ml/models/H2OBinaryModel.scala

```diff
@@ -0,0 +1,118 @@
+import ai.h2o.sparkling.backend.utils.{RestApiUtils, RestCommunication}
+import ai.h2o.sparkling.utils.ScalaUtils.withResource
+import ai.h2o.sparkling.utils.{ScalaUtils, SparkSessionUtils}
+import org.apache.hadoop.fs.Path
+import org.apache.spark.expose.Logging
+import water.AutoBuffer
+import water.api.schemas3.ModelsV3
+
+class H2OBinaryModel private (val modelId: String) extends HasBinaryModel with Logging with RestCommunication {
+
+  def write(path: String): Unit = {
+    val sc = SparkSessionUtils.active.sparkContext
+    val outputPath = new Path(path)
+    val fs = outputPath.getFileSystem(sc.hadoopConfiguration)
+    val qualifiedOutputPath = outputPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
+    withResource(fs.create(qualifiedOutputPath)) { out =>
+      Files.copy(new File(getBinaryModel().get.getAbsolutePath).toPath, out)
+    }
+    logInfo(s"Binary Model Saved to: $qualifiedOutputPath")
+  }
+}
+
+object H2OBinaryModel extends RestCommunication {
+
+  def binaryModelExists(modelId: String): Boolean = {
+    val conf = H2OContext.ensure().getConf
+    val endpoint = RestApiUtils.getClusterEndpoint(conf)
+    val models = query[ModelsV3](endpoint, "/3/Models", conf)
+    models.models.exists(_.model_id.name == modelId)
+  }
+
+  def createFromBinary(path: String): H2OBinaryModel = {
+    createFromBinary(path, None)
+  }
+
+  private def extractVersionFromModel(path: String): String = {
+    ScalaUtils.withResource(SparkSessionUtils.readHDFSFile(path)) { fis =>
+      ScalaUtils.withResource(new DataInputStream(fis)) { dis =>
+        // The number of 20 is chosen without any special meaning, but the purpose is the following:
+        // We need to read the version from binary model and instead just reading the whole binary file,
+        // we read the head.
+        // The head of the file looks like:
+        // 1 byte - persistence info
+        // 1 byte - magic number
+        // 1 up to 4 bytes - length of the string that follows ( H2O stores integer efficiently into smaller types when possible)
+        // The version itself. The version 30.30.10.10 consist of 11 characters => 11 bytes
+        val arr = new Array[Byte](20)
+        dis.readFully(arr)
+        val ab = new AutoBuffer(arr)
+        val persistenceInfo = ab.get1U() // Byte indicates persistence info
+        val magic = ab.get1U() // Magic byte
+        if (persistenceInfo != 0x1C || magic != 0xED) {
+          throw new RuntimeException("Invalid binary model")
+        }
+        ab.getStr
+      }
+    }
+  }
+
+  private def checkVersion(expected: String, modelVersion: String): Unit = {
+    if (expected != modelVersion) {
+      throw new IllegalArgumentException(s"""
+          | The binary model has been trained in H2O of version
+          | $modelVersion but you are currently running H2O version of $expected.
+          | To be able to use binary models, please make sure that Sparkling water/H2O is of the
```

Comment:
nit: Maybe, to split this sentence would help readability.

Reply:
Any suggestions please? For me, this sentence makes sense

Reply:
This was just suggestion to have shorter sentences. Maybe the second part could rephrased like this.:
```please make sure that running Sparkling Water/H2O-3 cluster and the loaded binary model correspond to the same H2O-3 version.

Reply:
Like it, thanks!